Respect HTML colors in labels

nyalldawson opened the pull request:
When enabled, this option will treat label contents as HTML, and
any FOREGROUND COLOR RELATED html formatting options will be
respected in the rendered label.

I repeat: ONLY HTML COLOR TAGS ARE RESPECTED. This is NOT a bug,
rather this feature has been designed as a "test of the waters"
with HTML formatting in labels, and accordingly only formatting
options which do not alter the font shape were considered.

On the plus side, it works correctly with ALL other label settings,
including shadows, buffers, curved labels, etc!

Sponsored by geoProRegio AG (www.geoproregio.ch)

Now some pretty pictures:

![image](https://user-images.githubusercontent.com/1829991/81377287-84359b80-9148-11ea-95f4-9d7917d62cf9.png)

![image](https://user-images.githubusercontent.com/1829991/81377294-88fa4f80-9148-11ea-9036-2a597b064bbc.png)

nyalldawson committed:
Add html flag

nyalldawson committed:
HTML formatting checkbox

nyalldawson committed:
Add method to extract blocks from HTML text for labeling

nyalldawson committed:
Adapt QgsPalLabeling.splitToLines to account for HTML formatting

nyalldawson committed:
[FEATURE] Allow use of color related HTML font tags in label text

When enabled, this option will treat label contents as HTML, and
any FOREGROUND COLOR RELATED html formatting options will be
respected in the rendered label.

I repeat: ONLY HTML COLOR TAGS ARE RESPECTED. This is NOT a bug,
rather this feature has been designed as a "test of the waters"
with HTML formatting in labels, and accordingly only formatting
options which do not alter the font shape were considered.

On the plus side, it works correctly with ALL other label settings,
including shadows, buffers, curved labels, etc!

Sponsored by geoProRegio AG (www.geoproregio.ch)

m-kuhn commented on src/core/qgsrendercontext.cpp:
```diff
@@ -98,7 +98,8 @@ QgsRenderContext &QgsRenderContext::operator=( const QgsRenderContext &rh )
   mCustomRenderingFlags = rh.mCustomRenderingFlags;
   setIsTemporal( rh.isTemporal() );
-  setTemporalRange( rh.temporalRange() );
+  if ( rh.isTemporal() )
```
Unrelated change?

nyalldawson committed:
Create custom QGIS classes for QgsTextDocument, QgsTextBlock, QgsTextCharacterFormat
etc

In the hope of eventually removing the QTextDocument dependancy for HTML
labels. We use this ONLY for parsing the HTML/css, and it's VERY
heavy for just these needs.

nyalldawson committed:
Sip says no

nyalldawson committed:
General cleanups, and add support for underline/overline/strikethrough formatting tags

nyalldawson committed:
Cleanups, vertical text handling

nyalldawson committed:
More cleanups, avoid multiple parsing of the same HTML content for a label

nyalldawson commented:
Now also supports underline, strikethrough, and overline:

![image](https://user-images.githubusercontent.com/1829991/81532413-ad079c00-93a7-11ea-8f76-2128307d9939.png)

NathanW2 commented:
;) hello html labels my old friend. Good to see you are back from the dead

On Mon, 11 May 2020, 4:51 pm Nyall Dawson, <notifications@github.com> wrote:

> Now also supports underline, strikethrough, and overline:
>
> [image: image]
> <https://user-images.githubusercontent.com/1829991/81532413-ad079c00-93a7-11ea-8f76-2128307d9939.png>
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/qgis/QGIS/pull/36283#issuecomment-626507629>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAC5FXG4HQKW7VNA7SWD6QDRQ6OATANCNFSM4M35SQNQ>
> .
>

m-kuhn reviewed (approved):
I have reviewed this with respect to

 - Basic code style sanity checks
 - Feature awesomeness
 - API looking reasonable
 - Tests are in place

And can give a +1 for all of that.

nyalldawson committed:
Code shuffle

nyalldawson committed:
Split massive qgstextrenderer.cpp/.h files into separate files per
contained class

and tighten includes to speed up recompilation

nyalldawson committed:
Expose some more classes to Python, but mark as unstable API

nyalldawson committed:
Use composition rather then inheritance for QgsTextBlock/QgsTextDocument

Allows us to expose these to python and hopefully fixes compilation issue on Travis

nyalldawson committed:
Build fixes

nyalldawson committed:
Don't try to reuse True value in enum, causes Python issue

nyalldawson committed:
Try to fix Travis build

nyalldawson committed:
Update QgsTextRenderer::textWidth/textHeight to handle HTML formatting

nyalldawson committed:
Fix scalebar text size calculation when HTML formatting is used

nyalldawson committed:
Disabled test run

nyalldawson committed:
Travis build fix

nyalldawson committed:
Remove outdated test

nyalldawson committed:
PyQGIS syntatic sugar, start on tests for QgsTextDocument

nyalldawson committed:
Complete tests for QgsTextDocument

nyalldawson committed:
More tests

nyalldawson committed:
Unit tests for QgsTextCharacterFormat

nyalldawson committed:
Fix build

nyalldawson committed:
More tests

nyalldawson committed:
More reference images

nyalldawson committed:
Fix F523 flake issues

nyalldawson commented:
Ok, this is ready for merge (if CI is happy)

nyalldawson committed:
Test masks

nyalldawson committed:
Sipify

m-kuhn commented:
Thanks for the test work

nyalldawson committed:
Fix custom widgets build

nyalldawson committed:
Fix doxygen warning

timlinux commented:
I wonder more generally what the use case of this is - are there any maps in the wild you can show that use mixed colours etc in the labels?

Just curious...more options are always nice :-)

andreasneumann commented:
> I wonder more generally what the use case of this is - are there any maps in the wild you can show that use mixed colours etc in the labels?
> 
> Just curious...more options are always nice :-)

Yes - consider maps that show the state of industrial devices or infrastructure - e.g. waste-water structures with pump, valves, different levels. Color and bold can express the state of different sensors.

nyalldawson committed:
Doxygen test fixes
